### PR TITLE
[Core] Ray auto detect nvidia Gpu with pynvml

### PR DESCRIPTION
**Pull request by jonathan-anyscale**
## Why are these changes needed?
We want to vendor `nvidia-ml-py` to Ray in order to serve several purposes:
- `nvidia_gpu` auto detection current implementation didn't work when Ray doesn't have root access (#28064). From several alternatives mentioned in the issue, all of them requires `libnvi-ml.so` which is nvidia driver C api. So, we use `pynvml` which is a python wrapper of this  `libnvi-ml.so` without any other dependencies. Additionally, we can remove GPUUtil from auto detection as well.
- Removing `gpustat` dependencies on Ray[default] by replacing it with `pynvml` which `gpustat` depends on as well (later PR)

## Related issue number
#28064

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mattip** [comment]
This seems like the right direction, and will close #35581 and a number of issues around GPU detection.

> Additionally, we can remove GPUUtil from auto detection as well.

Will this be in a follow-up PR?

File: python/requirements.txt [hunk 1]
```diff
@@ -59,5 +58,6 @@ typer
 fsspec
 pandas>=1.3
 pydantic<2
+pynvml>=11.5.0
```
Comment (wookayin) [inline python/requirements.txt:61]: [pynvml](https://github.com/gpuopenanalytics/pynvml) should be avoided, one should use [nvidia-ml-py](https://pypi.org/project/nvidia-ml-py/).

File: python/setup.py [hunk 2]
```diff
@@ -343,6 +342,7 @@ def get_packages(self):
         "pyyaml",
         "aiosignal",
         "frozenlist",
+        "pynvml",
```
Comment (wookayin) [inline python/setup.py:345]: here too

**Comment by jonathan-anyscale** [comment]
> This seems like the right direction, and will close #35581 and a number of issues around GPU detection.
> 
> > Additionally, we can remove GPUUtil from auto detection as well.
> 
> Will this be in a follow-up PR?

The auto detection will be in this PR as well as the change is not too big + part of using nvml library. Will update once I'm done with the changes.

File: python/requirements.txt (hunk 1 above)
Comment (jjyao) [reply python/requirements.txt:61]: @wookayin do you know the difference between these two?

Comment (wookayin) [reply python/requirements.txt:61]: `nvidia-ml-py` is the only official python NVML bindings maintained by NVIDIA. `pynvml` is not. I have no idea why this package exists, only adds a confusion. It adds some additional packages but the core nvml part is simply a copy, redistribution of the official pynvml. It's also not up-to-date (as of now, it doesn't support R535 drivers for instance, although the use here would be still backward-compatible), and used to behave *differently* than the official one and has caused some conflicts in the past.

Comment (jjyao) [reply python/requirements.txt:61]: @wookayin Thanks for the detailed explanation! 

> It's also not up-to-date (as of now, it doesn't support R535 drivers for instance, although the use here would be still backward-compatible)

Since we use pynvml to auto detect the number of gpus, will an old version of pynvml be able to detect newer version of gpus or users need to use the latest nvidia-ml-py.

Comment (wookayin) [reply python/requirements.txt:61]: That's true, for this purpose needed by ray, either will work. But other packages (including gpustat and nvitop to name a few) requires nvidia-ml-py, and if both packages (`pip install pynvml nvidia-ml-py`) are installed, which provides the same python package, then one will overwrite the other, which will break the `pynvml` package. In other words, installing `ray` shipping with the wrong dependency might break `gpustat` or `nvitop`. This should be avoided, so `nvidia-ml-py` should be used instead of `pynvml`.

Comment (jjyao) [reply python/requirements.txt:61]: Yea we will use `nvidia-ml-py` not `pynvml`. But instead of depending on `nvidia-ml-py` in setup.py, we are considering just copying this `pynvml.py` file into Ray codebase. Do you see a problem with this approach? Like will, let's say, nvidia releases a new gpu tomorrow and the existing `pynvml.py` won't be able to auto-detect the count.

Comment (wookayin) [reply python/requirements.txt:61]: NVIDIA's pynvml often introduces unexpected breaking changes -- but in the past they were mostly about the process information. It was more than once so I don't fully trust NVIDIA for the guaranteed backward compatibility. That being said, it's quite unlikely the API for # of gpus, or memory utilization will break, but there's no guarantee. I think bundling `pynvml` is also a good idea which will be more protective against potential breaking changes -- as long as maintaining it is not a problem.

Comment (jjyao) [reply python/requirements.txt:61]: Sg. I think then bundling `pynvml` is the way to go for us.

File: python/ray/_private/accelerators/nvidia_gpu.py [hunk 3]
```diff
@@ -53,55 +50,82 @@ def get_current_process_visible_accelerator_ids() -> Optional[List[str]]:
...
+                        )
+                        mig_uuid = ""
+                        if Version(driver_version) >= Version(MIG_UUID_DRIVER_VERSION):
+                            mig_uuid = pynvml.nvmlDeviceGetUUID(mig_handle)
+                        else:
+                            mig_uuid = (
+                                f"MIG-{pynvml.nvmlDeviceGetUUID(handle)}"
+                                f"/{pynvml.nvmlDeviceGetComputeInstanceId(mig_handle)}"
+                                f"/{pynvml.nvmlDeviceGetGpuInstanceId(mig_handle)}"
+                            )
```
Comment (XuehaiPan) [inline python/ray/_private/accelerators/nvidia_gpu.py:84]:
```suggestion
                            mig_uuid = f"MIG-{pynvml.nvmlDeviceGetUUID(handle)}"
```

It should be either:

```text
MIG-<MIG-UUID>
```

or

```text
MIG-GPU-<GPU-UUID>/<GI ID>/<CI ID>
```

Comment (XuehaiPan) [inline python/ray/_private/accelerators/nvidia_gpu.py:84]:
```suggestion
                            mig_uuid = (
                                f"MIG-GPU-{pynvml.nvmlDeviceGetUUID(handle)}"
                                f"/{pynvml.nvmlDeviceGetComputeInstanceId(mig_handle)}"
                                f"/{pynvml.nvmlDeviceGetGpuInstanceId(mig_handle)}"
                            )
```

The UUID for a MIG device is either:

```text
MIG-<MIG UUID>
```

or

```text
MIG-GPU-<GPU UUID>/<GI ID>/<CI ID>
```

Comment (jonathan-anyscale) [reply python/ray/_private/accelerators/nvidia_gpu.py:84]: https://docs.nvidia.com/datacenter/tesla/mig-user-guide/index.html#cuda-visible-devices
based on this, both gpu and mig uuid should've include `GPU-` and `MIG-`

Comment (XuehaiPan) [reply python/ray/_private/accelerators/nvidia_gpu.py:84]: Oh, sorry. You are right. The NVML API `nvmlDeviceGetUUID()` will return a UUID with the `GPU-` or `MIG-` prefix:

```python
nvmlDeviceGetUUID(gpu_handle)  # -> GPU-abcdef12-3456-7890-abcd-ef1234567890
nvmlDeviceGetUUID(mig_handle)  # -> MIG-a1b2c3d4-e5f6-7890-a1b2-c3d4e5f67890
```

The UUID for a MIG device should look like:

```text
MIG-a1b2c3d4-e5f6-7890-a1b2-c3d4e5f67890
MIG-GPU-abcdef12-3456-7890-abcd-ef1234567890/9/1
```

File: python/ray/_private/thirdparty/pynvml/__init__.py [hunk 4]
```diff
@@ -0,0 +1 @@
+from ray._private.thirdparty.pynvml.pynvml import *
```
Comment (wookayin) [inline python/ray/_private/thirdparty/pynvml/__init__.py:1]: Just a suggestion: It'd be good to keep track of which version of `nvidia-ml-py` is being imported (for future references) as metadata. `pynvml` does not have any sort of `__verion__` strings.

Comment (wookayin) [reply python/ray/_private/thirdparty/pynvml/__init__.py:1]: Also, I think this is fine with the current space, but just heads up: As in https://github.com/wookayin/gpustat/issues/153, this does not re-export private members starting with `_`. Private APIs of `pynvml` are not being used so it'd be fine :)

File: python/ray/_private/ray_constants.py [hunk 5]
```diff
@@ -480,3 +480,5 @@ def gcs_actor_scheduling_enabled():
 
 
 RAY_NODE_IP_FILENAME = "node_ip_address.json"
+
+RAY_ENABLE_MIG_DETECTION_ENV_VAR = "RAY_ENABLE_MIG_DETECTION"
```
Comment (jjyao) [inline python/ray/_private/ray_constants.py:484]: Let's do MIG support in a separate PR. For this PR, let's just remove gpustats and GPUtil and replace them with pynvml.

File: python/ray/_private/thirdparty/pynvml/__init__.py (hunk 4 above)
Comment (jjyao) [reply python/ray/_private/thirdparty/pynvml/__init__.py:1]: +1. We should track the version of `nvidia-ml-py` that we are vendoring here.

File: python/ray/_private/ray_constants.py (hunk 5 above)
Comment (jonathan-anyscale) [reply python/ray/_private/ray_constants.py:484]: ohh this is just to enable mig instance detection, for mig feature later, it will be a flag `RAY_ENABLE_MIG_ENV_VAR`

File: python/ray/_private/thirdparty/pynvml/__init__.py (hunk 4 above)
Comment (jonathan-anyscale) [reply python/ray/_private/thirdparty/pynvml/__init__.py:1]: sounds good for the version.

@wookayin for private APIs, yes for now I dont think we will use it so keeping it that way + change in case we need later

File: python/ray/_private/ray_constants.py (hunk 5 above)
Comment (jjyao) [reply python/ray/_private/ray_constants.py:484]: Yea I'm suggesting we remove all mig related changes from this PR and just focus on removing gpustat and GPUtil dependencies. This should already be a big enough change for this PR since you also need to fix dashboard.

Comment (jonathan-anyscale) [reply python/ray/_private/ray_constants.py:484]: I planned to split the change with auto detect and remove gpustat PRs instead since one cover entirely core while other is entirely dashboard
https://github.com/ray-project/ray/pull/41044 for the gpustat pr

File: python/ray/_private/accelerators/nvidia_gpu.py [hunk 6]
```diff
@@ -53,55 +50,82 @@ def get_current_process_visible_accelerator_ids() -> Optional[List[str]]:
...
+                            handle, mig_index
+                        )
+                        mig_uuid = ""
+                        if Version(driver_version) >= Version(MIG_UUID_DRIVER_VERSION):
+                            mig_uuid = pynvml.nvmlDeviceGetUUID(mig_handle)
+                        else:
+                            mig_uuid = (
+                                f"MIG-{pynvml.nvmlDeviceGetUUID(handle)}"
+                                f"/{pynvml.nvmlDeviceGetComputeInstanceId(mig_handle)}"
+                                f"/{pynvml.nvmlDeviceGetGpuInstanceId(mig_handle)}"
```
Comment (XuehaiPan) [inline python/ray/_private/accelerators/nvidia_gpu.py:83]:
```suggestion
                                f"/{pynvml.nvmlDeviceGetGpuInstanceId(mig_handle)}"
                                f"/{pynvml.nvmlDeviceGetComputeInstanceId(mig_handle)}"
```

File: python/ray/tests/accelerators/test_nvidia_gpu.py [hunk 7]
```diff
@@ -1,43 +1,209 @@
...
+
+    def nvmlDeviceGetHandleByIndex(self, index):
+        return self._mock_data[index]
+
+    def nvmlDeviceGetName(self, handle):
+        return handle.get("name", "")
+
+    def nvmlDeviceGetMaxMigDeviceCount(self, handle):
+        if "mig_devices" in handle:
+            return len(handle["mig_devices"])
```
Comment (XuehaiPan) [inline python/ray/tests/accelerators/test_nvidia_gpu.py:91]: `nvmlDeviceGetMaxMigDeviceCount()` will return either 7 or 0 on MIG-supported devices, or raise `NVMLError_NotSupported` error for unsupported devices.

File: python/ray/tests/accelerators/test_nvidia_gpu.py [hunk 8]
```diff
@@ -1,43 +1,209 @@
...
+        else:
+            raise pynvml.NVMLError_NotSupported
+
+    def nvmlDeviceGetMigDeviceHandleByIndex(self, handle, mig_index):
+        return handle["mig_devices"][mig_index]
```
Comment (XuehaiPan) [inline python/ray/tests/accelerators/test_nvidia_gpu.py:96]:
```suggestion
    def nvmlDeviceGetMigDeviceHandleByIndex(self, handle, mig_index):
        try:
            return handle["mig_devices"][mig_index]
        except IndexError:
            raise pynvml.NVMLError_NotFound
```

File: python/ray/_private/accelerators/nvidia_gpu.py [hunk 9]
```diff
@@ -53,58 +45,29 @@ def get_current_process_visible_accelerator_ids() -> Optional[List[str]]:
...
-            logger.exception("Could not parse gpu information.")
-        return None
+            pynvml.nvmlInit()
+        except pynvml.NVMLError:
+            return None  # pynvml init failed
+        device_count = pynvml.nvmlDeviceGetCount()
+        cuda_devices_names = []
+        for index in range(device_count):
+            handle = pynvml.nvmlDeviceGetHandleByIndex(index)
+            cuda_devices_names.append(pynvml.nvmlDeviceGetName(handle))
```
Comment (jjyao) [inline python/ray/_private/accelerators/nvidia_gpu.py:66]: Seems you just use the first cuda_devices_names, so don't need this for loop?

**Comment by jonathan-anyscale** [comment]
As per https://github.com/ray-project/ray/pull/41020#discussion_r1388207822
We will move mig detection support as follow up PR

**Review by jjyao — APPROVED** [review]
@wookayin @XuehaiPan it'd be great if you can also review our usage of pynvml since you are experts here :)

File: python/ray/_private/accelerators/nvidia_gpu.py (hunk 9 above)
Comment (jonathan-anyscale) [reply python/ray/_private/accelerators/nvidia_gpu.py:66]: I just follow the old implementation which use GPUtil, but want to confirm we just want to get the first device name then?

Comment (jjyao) [reply python/ray/_private/accelerators/nvidia_gpu.py:66]: From your code, you only used the first one?

Comment (jonathan-anyscale) [reply python/ray/_private/accelerators/nvidia_gpu.py:66]: Right, I'll remove the loop. Seems like the old implementation also only return the first device name despite there are multiple gpus
